sessionctx/sessionstates/session_token.go

```diff
@@ -0,0 +1,338 @@
+	// However, we need to be tolerant of these problems:
+	// - The `tokenLifetime` may change between TiDB versions, so we can't check `token.SignTime.Add(tokenLifetime).Equal(token.ExpireTime)`
+	// - There may exist time bias between TiDB instances, so we can't check `now.After(token.SignTime)`
+	if token.SignTime.Add(tokenLifetime).Before(now) {
+		return ErrCannotMigrateSession.GenWithStackByArgs("token lifetime is too long", token.SignTime.String())
+	}
+	if !strings.EqualFold(username, token.Username) {
+		return ErrCannotMigrateSession.GenWithStackByArgs("username does not match", username, token.Username)
+	}
+	return nil
+}
+
+// SetKeyPath sets the path of key.pem and force load the certificate again.
+func SetKeyPath(keyPath string) {
+	globalSigningCert.setKeyPath(keyPath)
+}
+
+// SetCertPath sets the path of key.pem and force load the certificate again.
+func SetCertPath(certPath string) {
+	globalSigningCert.setCertPath(certPath)
+}
+
+// ReloadSigningCert is used to load the certificate periodically in a separate goroutine.
+// It's impossible to know when the old certificate should expire without this goroutine:
+// - If the certificate is rotated a minute ago, the old certificate should be still valid for a while.
+// - If the certificate is rotated a month ago, the old certificate should expire for safety.
+func ReloadSigningCert() {
+	globalSigningCert.lockAndLoad()
+}
+
+var globalSigningCert signingCert
+
+// signingCert represents the parsed certificate used for token-based auth.
+type signingCert struct {
+	sync.RWMutex
+	certPath string
+	keyPath  string
+	// The cert file may happen to be rotated between signing and checking, so we keep the old cert for a while.
+	// certs contain all the certificates that are not expired yet.
+	certs []*certInfo
+}
+
+type certInfo struct {
+	cert       *x509.Certificate
+	privKey    crypto.PrivateKey
+	expireTime time.Time
+}
+
+// We cannot guarantee that the cert and key paths are set at the same time because they are set through system variables.
+func (sc *signingCert) setCertPath(certPath string) {
+	sc.Lock()
+	// Just in case of repeatedly loading global variables, we check the path to avoid useless loading.
+	if certPath != sc.certPath {
+		sc.certPath = certPath
+		// It may fail expectedly because the key path is not set yet.
+		sc.checkAndLoadCert()
+	}
+	sc.Unlock()
+}
+
+func (sc *signingCert) setKeyPath(keyPath string) {
+	sc.Lock()
+	if keyPath != sc.keyPath {
+		sc.keyPath = keyPath
+		// It may fail expectedly because the cert path is not set yet.
+		sc.checkAndLoadCert()
+	}
+	sc.Unlock()
+}
+
+func (sc *signingCert) lockAndLoad() {
+	sc.Lock()
+	sc.checkAndLoadCert()
+	sc.Unlock()
+}
+
+func (sc *signingCert) checkAndLoadCert() {
+	if len(sc.certPath) == 0 || len(sc.keyPath) == 0 {
+		return
+	}
+	if err := sc.loadCert(); err != nil {
+		logutil.BgLogger().Warn("loading signing cert failed",
```

Comment:
In the future we need to add this error to the metrics(grafana), but not in this PR.